### PR TITLE
backend, net: Support compression protocol

### LINKED ISSUE #361: Support compressed MySQL protocol
## Development Task

TiDB has already supported compressed MySQL protocol in https://github.com/pingcap/tidb/pull/36780, so TiProxy should also support it.

This helps with:
- Performance: Especially data intensive work over low bandwidth connections
- Cost: This can reduce cost when bandwidth is payed for, e.g. in Cloud environments.

### PR DESCRIPTION
**Pull request by djshow832**
### What problem does this PR solve?

Issue Number: close #361

Problem Summary:
- The compression protocol may help to ease the network bandwidth bottleneck. The protocol spec: https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_basic_compression_packet.html
- Now we have multiple protocols in `PacketIO`: proxy-protocol, TLS, compression, and uncompressed packets. Too many branches in one class make the code harder to maintain.

What is changed and how it works:
- Support the compression protocol in `compress.go`.
- The 3 protocols can be treated as layered protocols: packet -> compression(optional) -> TLS(optional) -> proxy-protocol(optional) -> raw connection. Each layer doesn't care about what the underlying protocol really is. So I decouple the protocols into:
  - `PacketIO` still handles MySQL packet reading and writing
  - (optional)`compressReadWriter` handles compression protocol
  - (optional)`tlsReadWriter` handles TLS
  - (optional)`proxyReadWriter` handles proxy-protocol
  - `basicReadWriter` wraps the raw connection and handles basic connection features like recording the traffic, checking if the peer is alive

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Tested with:
- MySQL client 8.0.33 -> TiProxy -> MySQL server 8.0.33
- MySQL connector/J 8.01 -> TiProxy -> MySQL server 8.0.33

Notable changes

- [ ] Has configuration change
- [ ] Has HTTP API interfaces change
- [ ] Has tiproxyctl change
- [ ] Other user behavior changes

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/proxy/net/compress_test.go [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+
+func checkData(t *testing.T, data []byte, b byte) {
+	for i := range data {
+		if data[i] != b {
+			require.Fail(t, fmt.Sprintf("expected %c, but got %c", b, data[i]))
+		}
+	}
+}
+
+func checkByteNum(t *testing.T, diff uint64, size int) {
```
Comment (xhebox) [inline pkg/proxy/net/compress_test.go:268]: Maybe `checkWrittenByteSize`? I feel a little bit confused.

**Review by xhebox — COMMENTED** [review]
LGTM

File: pkg/proxy/net/proxy.go [hunk 2]
```diff
@@ -6,13 +6,100 @@ package net
...
+		// We don't know whether the client has enabled proxy protocol.
+		// If it doesn't, reading data of len(MagicV2) may block forever.
+		header, err := prw.Peek(4)
+		if err != nil {
+			return 0, errors.Wrap(ErrReadConn, err)
+		}
+		if bytes.Equal(header[:], proxyprotocol.MagicV2[:4]) {
+			if _, err = prw.Discard(4); err != nil {
+				return 0, err
+			}
```
[the remark this answers is not in the document]
Comment (djshow832) [reply pkg/proxy/net/proxy.go:68]: No, they matched: `if bytes.Equal(header[:], proxyprotocol.MagicV2[:4])`.
The error is thrown from `Discard`, which means it's a connection error.

[the remark this answers is not in the document]
Comment (xhebox) [reply pkg/proxy/net/proxy.go:68]: I mean, if `proxyprotocol.MagicV2[4:]` mismatched. That is, if `prw.parseProxyV2()` returned with `proxyHeader = nil && err == nil`.

[the remark this answers is not in the document]
Comment (djshow832) [reply pkg/proxy/net/proxy.go]: Addressed.

**Review by xhebox — APPROVED** [review]

**Pull request merged**